alf/experience_replayers/replay_buffer.py

```diff
@@ -175,28 +175,66 @@ def clear(self):
+                corresponding to one single actor. If None, batch_size must
+                be num_environments. If not None, its shape should be
+                [batch_size]. We assume there is no duplicate ids in `env_id`.
+
         Returns:
             Returns all the items currently in the buffer. The shapes of the
             tensors are [B, T, ...] where B=num_environments, T=current_size
         Raises:
             tf.errors.InvalidArgumentError: if the current_size is not same for
                 all the environments
         """
-        size = tf.reduce_min(self._current_size)
-        max_size = tf.reduce_max(self._current_size)
-        tf.Assert(size == max_size, [
-            "Not all environment have the same size. min_size:", size,
-            "max_size:", max_size
+        if env_id is None:
+            buffer = tf.nest.map_structure(lambda buf: buf.value(),
+                                           self._buffer)
+            current_size = self._current_size.value()
+            current_pos = self._current_pos.value()
+        else:
+            buffer = tf.nest.map_structure(
+                lambda buf: tf.gather_nd(buf.value(), env_id), self._buffer)
+            current_size = tf.nest.map_structure(
+                lambda buf: tf.gather_nd(buf.value(), env_id),
+                self._current_size)
+            current_pos = tf.nest.map_structure(
+                lambda buf: tf.gather_nd(buf.value(), env_id),
+                self._current_pos)
+
+        size = tf.reduce_min(current_size)
+        max_size = tf.reduce_max(current_size)
+        tf.Assert(
+            size == max_size,
+            ["Not all environment have the same size: ",
+             str(current_size)])
+        pos = tf.reduce_min(current_pos)
+        max_pos = tf.reduce_max(current_pos)
+        tf.Assert(pos == max_pos, [
+            "Not all environment have the same current_pos. min_pos: ",
+            str(current_pos)
         ])
 
-        if size == self._max_length:
-            return tf.nest.map_structure(lambda buf: buf.value(), self._buffer)
+        if pos == 0 and size == self._max_length:
+            # pull everything from the buffer in order
+            return buffer
+        elif pos + size - 1 > self._max_length - 1:
+            # pull from _current_pos to the end of the buffer and then from
+            # the beginning
+            return tf.nest.map_structure(
+                lambda buf: tf.concat([
+                    buf[:, pos:, ...], buf[:, :pos + size - self._max_length,
+                                           ...]
+                ],
+                                      axis=1), buffer)
         else:
+            # pull directly, not going over the end of the buffer
             return tf.nest.map_structure(lambda buf: buf[:, :size, ...],
-                                         self._buffer)
+                                         buffer)
```

Comment:
```python
buf[:, pos:pos+size, ...]
```
Test cases do not catch it.

Reply:
`pos` is the ending position of the current buffer.  This line is correct, and is protected by the test line:170.

Since we only clear_all, when buffer is half full, we always read from 0th element in buf.  In this case, pos == size.

I haven't implemented anything that would lead to buffer beginning from an element that's not the 0th.  In order to cause that, we'll need to implement buffer.clear_n().  If you think this can be useful, I can add that, otherwise, I suggest we don't add unnecessary code?